syntax = "proto3";

package theo.stt;

service STTWorker {
  // Transcricao batch (arquivo completo)
  rpc TranscribeFile (TranscribeFileRequest) returns (TranscribeFileResponse);

  // Transcricao streaming (bidirecional)
  rpc TranscribeStream (stream AudioFrame) returns (stream TranscriptEvent);

  // Cancelamento de request/sessao
  rpc Cancel (CancelRequest) returns (CancelResponse);

  // Health check (unario)
  rpc Health (HealthRequest) returns (HealthResponse);
}

// --- Mensagens de TranscribeFile ---

message TranscribeFileRequest {
  string request_id = 1;
  bytes audio_data = 2;
  string language = 3;           // ISO 639-1 ou "auto" ou "mixed"
  string response_format = 4;    // json, verbose_json, text, srt, vtt
  float temperature = 5;
  repeated string timestamp_granularities = 6;  // "segment", "word"
  string initial_prompt = 7;
  repeated string hot_words = 8;
  string task = 9;               // "transcribe" ou "translate"
}

message TranscribeFileResponse {
  string text = 1;
  string language = 2;
  float duration = 3;
  repeated Segment segments = 4;
  repeated Word words = 5;
}

// --- Mensagens de TranscribeStream ---

message AudioFrame {
  string session_id = 1;
  bytes data = 2;               // PCM 16-bit 16kHz mono (ja preprocessado)
  bool is_last = 3;             // Sinaliza fim do stream
  string initial_prompt = 4;    // Contexto do segmento anterior
  repeated string hot_words = 5;
}

message TranscriptEvent {
  string session_id = 1;
  string event_type = 2;        // "partial" ou "final"
  string text = 3;
  int32 segment_id = 4;
  int64 start_ms = 5;
  int64 end_ms = 6;
  string language = 7;
  float confidence = 8;
  repeated Word words = 9;
}

// --- Mensagens compartilhadas ---

message Segment {
  int32 id = 1;
  float start = 2;
  float end = 3;
  string text = 4;
  float avg_logprob = 5;
  float no_speech_prob = 6;
  float compression_ratio = 7;
}

message Word {
  string word = 1;
  float start = 2;
  float end = 3;
  float probability = 4;
}

// --- Cancelamento ---

message CancelRequest {
  string request_id = 1;
  string session_id = 2;        // Para streaming
}

message CancelResponse {
  bool acknowledged = 1;
}

// --- Health ---

message HealthRequest {}

message HealthResponse {
  string status = 1;            // "ok", "loading", "error"
  string model_name = 2;
  string engine = 3;
  map<string, string> metadata = 4;
}
